src/tools/wasm-reduce.cpp

```diff
@@ -842,16 +838,36 @@ struct Reducer
     if (it != module->elementSegments.end()) {
       first = it->get()->data[0];
     }
-
     // try to reduce to first function. first, shrink segment elements.
     // while we are shrinking successfully, keep going exponentially.
     bool shrank = false;
     for (auto& segment : module->elementSegments) {
       shrank = shrinkByReduction(segment.get(), 100);
     }
     // the "opposite" of shrinking: copy a 'zero' element
+    if (!first) {
+      return;
+    }
+
     for (auto& segment : module->elementSegments) {
-      reduceByZeroing(segment.get(), first, 100, shrank);
+      reduceByZeroing(
+        segment.get(),
+        first,
+        [&](Expression* entry) {
+          if (entry->is<RefNull>()) {
+            // we don't need to replace a ref.null
+            return true;
+          } else if (first->is<RefNull>()) {
+            return false;
+          } else {
+            // Both are ref.func
+            auto* f = first->cast<RefFunc>();
+            auto* e = entry->cast<RefFunc>();
+            return f->func == e->func && f->type == e->type;
+          }
+        },
```

Comment:
@kripken Please take a look at this. I restored the logic for finding a `zero` as it was needed for non-reftypes modules, but the part where it needs to check if an element is equal to the `zero` has become bigger than it would seem necessary. Just want to make sure it's not doing unnecessary work.

Reply:
I think it looks good with the suggested changes.